### PR TITLE
QgsTask::waitForFinished without event loop

### PR DESCRIPTION
**Pull request by m-kuhn**
because the old approach with using signals, using deleteLater and using
an internal event loop is a very good recipe for big troubles.

## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgstaskmanager.cpp [hunk 1]
```diff
@@ -142,26 +144,19 @@ QList<QgsMapLayer *> QgsTask::dependentLayers() const
...
+  if ( mOverallStatus == Complete || mOverallStatus == Terminated )
   {
-    timer.start( timeout );
-    connect( &timer, &QTimer::timeout, [&rv]() { rv = false; } );
-    connect( &timer, &QTimer::timeout, &loop, &QEventLoop::quit );
+    rv = true;
+  }
+  else
+  {
+    if ( timeout == -1 )
```
Comment (rldhont) [inline src/core/qgstaskmanager.cpp]: If `timeout` is an unsigned long, I suppose that it will never be equal to `-1`.

**Comment by rldhont** [comment]
Thanks @m-kuhn to work on segfault!

File: src/core/qgstaskmanager.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/core/qgstaskmanager.cpp]: Oops, I just pushed an updated version but apparently only changed the signature. Good catch.

Comment (m-kuhn) [reply src/core/qgstaskmanager.cpp]: Ah, QtCreator had not saved the file...

**Comment by m-kuhn** [comment]
@rldhont do you have an idea what's going wrong here?

```
======================================================================
FAIL: test_project_wms (__main__.TestQgsServerWMS)
Test some WMS request
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/qgis/QGIS/tests/src/python/test_qgsserver_wms.py", line 107, in test_project_wms
    'wms_getfeatureinfo_filter')
  File "/home/travis/build/qgis/QGIS/tests/src/python/test_qgsserver_wms.py", line 65, in wms_request_compare
    self.assertXMLEqual(response, expected, msg="request %s failed.\nQuery: %s\nExpected file: %s\nResponse:\n%s" % (query_string, request, reference_path, response.decode('utf-8')))
  File "/home/travis/build/qgis/QGIS/tests/src/python/test_qgsserver.py", line 69, in assertXMLEqual
    self.assertEqual(len(expected_lines), len(response_lines), "Expected and response have different number of lines!\n{}".format(msg))
AssertionError: 15 != 16 : Expected and response have different number of lines!
request https://www.qgis.org/?MAP=/home/travis/build/qgis/QGIS/tests/testdata/qgis_server/test_project.qgs&SERVICE=WMS&VERSION=1.3&REQUEST=GetFeatureInfo&layers=testlayer%20%C3%A8%C3%A9&INFO_FORMAT=text%2Fxml&width=600&height=400&srs=EPSG%3A3857&query_layers=testlayer%20%C3%A8%C3%A9&FEATURE_COUNT=10&FILTER=testlayer%20%C3%A8%C3%A9%3A%22NAME%22%20%3D%20%27two%27 failed.
Query: GetFeatureInfo
Expected file: /home/travis/build/qgis/QGIS/tests/testdata/qgis_server/wms_getfeatureinfo_filter.txt
Response:
*****
Content-Type: text/xml; charset=utf-8
<GetFeatureInfoResponse>
 <BoundingBox CRS="EPSG:3857" minx="913214.67407005" maxx="913214.67407005" miny="5606017.87425818" maxy="5606017.87425818"/>
 <Layer name="testlayer èé">
  <Feature id="2">
   <Attribute name="id" value="2"/>
   <Attribute name="name" value="two"/>
   <Attribute name="utf8nameè" value="two àò"/>
   <Attribute name="orig_ogc_fid" value="1"/>
   <BoundingBox CRS="EPSG:3857" minx="913214.6741" maxx="913214.6741" miny="5606017.8743" maxy="5606017.8743"/>
   <Attribute name="geometry" value="Point (913214.6741 5606017.8743)" type="derived"/>
  </Feature>
 </Layer>
</GetFeatureInfoResponse>
```

**Comment by rldhont** [comment]
I don't know why, but the difference is an empty line between head and content

**Comment by m-kuhn** [comment]
I don't know how my change could have triggered that, do you? What about the mismatch in feature id?

**Comment by rldhont** [comment]
something about re-affecting featureid based on count ?

**Comment by m-kuhn** [comment]
Can you have a look at that?

**Comment by rldhont** [comment]
I'll check it.

**Comment by elpaso** [comment]
@m-kuhn that test has been randomly failing for a few days. Feel free to disable it, I'll not complain ;)

**Comment by m-kuhn** [comment]
Thanks for the feedback @elpaso it's now disabled.
It would of course be great if someone still could have a look and possibly make it less prone by ignoring the line-count and either switch to a provider that generates stable feature ids or also ignore this line.

**Pull request merged**